File: Numeros Pseudo-Aleatorios/Lehmer/Lehmer.scala
/*
PRNG .......: Lehmer (Metodo Congruencial Linear Multiplicativo)
Ano ........: 1951
Autor ......: Derrick Henry "Dick" Lehmer
*/

object Main {

  def main(args: Array[String]): Unit = {
    
    object PRNG {
      def Lehmer(seed: Long): Long = {
        var currentSeed = seed
        currentSeed = (16807 * currentSeed) % 2147483647
        currentSeed & 0x7FFFFFFF // Manter apenas os bits positivos
      }
    }

    println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)")
    println("PSEUDORANDOM NUMBER GENERATOR (PRNG)")
    println("PADRAO: Lehmer")
    println("==========================================\n")
    println("    Semente: 1234        Semente: TIME")
    println("    -------------        -------------")

    var Xn1: Long = 1234
    var Xn2: Long = System.currentTimeMillis() / 1000

    for (i <- 1 to 10) {
      Xn1 = PRNG.Lehmer(Xn1)
      Xn2 = PRNG.Lehmer(Xn2)
      println(f"${Xn1}%17d        ${Xn2}%13d")
    }
    
  }

}

File: Numeros Pseudo-Aleatorios/Lehmer/Lehmer.c
/* 
 PRNG .......: Lehmer (Metodo Congruencial Linear Multiplicativo)
 Ano ........: 1951
 Autor ......: Derrick Henry "Dick" Lehmer
*/

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

long Lehmer(long *seed) {
    *seed = (16807 * (*seed)) % 2147483647;
    return *seed;
}

int main(void) {

    long Xn1, Xn2;
    int i;

    Xn1 = 1234;
    Xn2 = (long)time(NULL);

    printf("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)\n");
    printf("PSEUDORANDOM NUMBER GENERATOR (PRNG)\n");
    printf("PADRAO: Lehmer\n");
    printf("==========================================\n\n");
    printf("    Semente: 1234        Semente: TIME\n");
    printf("    -------------        -------------\n");

    for (i = 1; i <= 10; i++) {
        printf("%17ld        %13ld\n", Lehmer(&Xn1), Lehmer(&Xn2));
    }

    return 0;
}
